### ISSUE TITLE
Feature request: snap to grid

### ISSUE THREAD

**Issue by fluffy-critter** [25A0/Quadtastic#23]
I generally design my sprites to conform to an 8x8 tile grid, and I would love it if I could have the quads snap to this grid, both when dragging them manually and when doing the magic wand tool. This would make it a lot easier to author padding into sprites directly, useful for frame-by-frame animation in many cases.

Similarly, being able to add top/left/bottom/right padding to the quad, with those attributes added to the quad object after the fact when libquadtastic parses the file, e.g.

        if libquadtastic.is_quad(v) then
          t[k] = love.graphics.newQuad(v.x, v.y, v.w, v.h, width, height)
          t[k].pad = v.pad
        elseif ...

This is enough of a feature that I'd be fine with taking a stab at writing it myself if you don't want to do it. :)

**Comment by 25A0** [comment]
I like your ideas for grid alignment :)

Some additional thoughts:

 - I had a quick look at how Gimp and Aseprite handle grids. In Gimp, things only snap to the grid when your mouse is kind of close to the grid lines or grid points, while in Aseprite, things will always and reliably snap to the grid. I much prefer Aseprite's approach. How about you?
 - I think it might be useful if snapping to the grid can be easily toggled by holding one of the modifier keys, e.g. <kbd>alt</kbd>.
 - There should be a new menu called 'Grid' (maybe in the 'Edit' menu? I'm not good at organizing menus) with the following entries:
   - a checkbox "Always snap to grid". If all your sprites are grid-aligned, you can enable this option so that you don't need to press that modifier key constantly. If this option is enabled, pressing the modifier key will temporarily disable snapping to the grid.
   - a menu called 'Grid size' to define the grid size. It has a number of entries for presets (e.g. 4x4, 8x8, 12x12, 16x16, 20x20, 24x24, 32x32) as well as an entry called 'Custom size...' that opens a small dialog which lets you define an arbitrary size. It might also be cool to keep track of recently used custom sizes, similar to how recently opened files are remembered.
 - The currently used grid size should be saved in the settings so that it is restored when you close and re-open Quadtastic
 - The texture in the background, which currently shows a 4x4 grid, should reflect the current grid size.

What do you think?

I'll happily implement this myself, but I'm quite busy until mid-september, so progress might be slow for the next two weeks or so. If you need this feature sooner, feel free to implement what you need (e.g. just snapping to an 8x8 grid, without all the interface stuff), and let me know when you need help with anything :)

Could you elaborate on the way you envision padding to work? Should Quadtastic detect padding automatically or would you like to be able to define a padding attribute manually for each quad?

**Comment by fluffy-critter** [comment]
Those are all very good thoughts! In the specific case of what I want out of this I'd want it to always snap to the grid (I can't even think of a time I'd personally want to override it) but a modifier key as a temporary toggle is of course a good idea. And I like your idea for the menu. 

Padding I was thinking would be an additional border that's associated with the sprite although I haven't put any thought into how the UI would look. It was sort of a "would be nice" musing. Actually something that would be even more useful (and probably easier to implement) is just having an offset property (x,y) that indicates where the natural reference position of the sprite is, and that would be indicated by e.g. a red dot that you can place and drag around on each quad. One complicating thing is that sometimes one might want the reference dot to be outside the quad in question (like for aligning multiple separate sprite layers together). A reference position dot would actually be more useful to me than a grid snap, come to think of it...

**Comment by 25A0** [comment]
> Actually something that would be even more useful (and probably easier to implement) is just having an offset property (x,y) that indicates where the natural reference position of the sprite is, and that would be indicated by e.g. a red dot that you can place and drag around on each quad. One complicating thing is that sometimes one might want the reference dot to be outside the quad in question (like for aligning multiple separate sprite layers together)

Just to make sure that I understand what you mean, is this what you have in mind?
![screen shot 2017-08-28 at 21 34 57](https://user-images.githubusercontent.com/6221922/29790095-c6f20a7a-8c38-11e7-9d78-d1efa231b5d6.png)
(excuse my shitty art :D)
The red dots would be these reference positions, and by aligning them you get the image on the right.

If this is what you had in mind, it seems to me that you could achieve the same by having grid-aligned quads of the same size around each sprite, right? If you then draw all these sprites at the same location, the sprites would align correctly.
![screen shot 2017-08-28 at 21 35 27](https://user-images.githubusercontent.com/6221922/29790106-d6acf72c-8c38-11e7-96d8-028e2b19692d.png)

**Comment by fluffy-critter** [comment]
Something like that! Which is why I was thinking grid-aligned quads originally, since that handles the thing fine. But I was also thinking the reference position might not necessarily be based on the top-left corner; there's a lot of situations where it makes more sense to have it placed on the character's foot, for example - placing the sprite based on where they are in the world, rather than where an arbitrary point to the up-and-left of their head happens to be :)

**Comment by 25A0** [comment]
I see :)
What if we turned this into two separate features?

One is general grid alignment, because I think that this is just generally useful for a number of different workflows.

The other is annotating quads with additional details like points, lines, or rectangles. I'm a bit hesitant to add something like a reference position explicitly. Instead I'm thinking about allowing users to add arbitrary attributes to their quads. This could be useful for a whole range of things. A few examples:

 - general cases like the reference position that you'd like to have
 - specifying the position of chimneys on sprites of houses, so that you know where to generate smoke particles
 - tracking the position of specific points (like a hand) in a frame-by-frame animation
 - aligning the base of a windmill with the blades
 - specifying positions on a sofa sprite where people can sit
 - defining (rectangular) hot spots in background sprites of an adventure game

All these things would usually end up as hard-coded numbers in the code. I think it would fit the spirit of Quadtastic if there was a way to specify these attributes on the spritesheet itself.

You could have as many attributes per quad as you like, and you'd be able to name them, too, just like quads. This would allow users to define what they need; they don't depend on me giving them options that happen to match their workflow. While you would like to define a reference position, others might want to rotate their sprites and might want to define a center of rotation for that.

I'm curious to hear what you think :) I don't have a lot of game dev experience, so while such a feature sounds useful in my head, there may certainly be valid reasons why this would not be useful in practice :)

**Comment by fluffy-critter** [comment]
They are definitely separate features yeah. I only brought up the reference point on this issue because it occurred to me as a potential alternative. I'd totally be fine with splitting that out into a separate issue. 

In any case, having arbitrary annotations would be great, if you can think of a good UI for managing it. :)

**Comment by 25A0** [comment]
> In any case, having arbitrary annotations would be great, if you can think of a good UI for managing it. :)

Yeah, I have a few ideas how to make these convenient to use, but it's not trivial... 

Thanks for your input and feedback! I'll get started with implementing grid alignment as soon as I find some time :)

**Comment by 25A0** [comment]
I've started to implement the grid feature, and will push commits to branch [`snap-to-grid`](https://github.com/25A0/Quadtastic/tree/snap-to-grid).

Progress so far:
 - [x] When using the create or wand tool, the positions and dimensions of quads are aligned with a grid
 - [ ] When using the resize and move tool, the size and position of changed quads should snap to the grid in some way.
 - [ ] When using the palette tool, the created quads should snap to the grid
 - [ ] A new menu called "Grid" in the "Edit" menu is added to configure the grid 
   - [x] A checkbox "Always snap to grid" defines whether quads should always snap to the grid
   - [ ] A sub-menu lets the user pick the grid dimensions
     - [x] A number of presets (4x4, 8x8, 12x12, 16x16, 20x20, 24x24, 32x32) are available
     - [ ] The user can define custom grid dimensions in a small dialog
     - [ ] The 10 most recent custom dimensions are stored in the settings, and displayed alongside the presets for convenience
 - [x] A modifier key (currently <kbd>alt</kbd>) can be used to temporarily toggle whether quads should snap to the grid - [x] The grid settings are stored in the application settings so that they are restored when Quadtastic is closed and re-opened
 - [x] The texture in the background of the sprite sheet reflects the current grid dimensions

**Comment by fluffy-critter** [comment]
Holy cow you work fast. Thank you so much for doing this!

**Comment by 25A0** [comment]
> Thank you so much for doing this!
Np! I enjoy it :)

As far as I'm concerned, grid snapping is now feature-complete, but the code is still on the `snap-to-grid` branch. If there are any changes to this feature that you'd like to see, let me know :)

There is one thing that I haven't quite figured out yet.
![screen shot 2017-09-03 at 01 15 59](https://user-images.githubusercontent.com/6221922/29999356-81325d58-9045-11e7-8809-dcc61ebbc632.png)
This is a set of glyphs in a 5x7 grid. For all but the last two glyphs, the wand tool creates one quad for each glyph while grid snapping is enabled. The last two glyphs, however, touch in one spot, so that the wand tool detects a continuous area, and creates a single quad that covers both glyphs.
I could solve this by restricting the wand tool to only create quads as large as one tile (so 5x7 in this case) when grid snapping is enabled.

**Comment by fluffy-critter** [comment]
Thanks, I’ll check it out!

In the case of grid spilling, that’s actually the behavior I’d expect, since I’m using this for making sprites that are often multiple “cells” large. Fonts I usually make with bmfont or some other tool that provides better built-in support for LÖVE’s text rendering anyway.

**Comment by fluffy-critter** [comment]
Hm, I tried running this branch and I got:

```Error: /Quadtastic.lua:73: attempt to index field 'grid' (a nil value)
stack traceback:
	/Quadtastic.lua:73: in function 'get_default_settings'
	/Quadtastic.lua:138: in main chunk
	[C]: in function 'require'
	main.lua:11: in main chunk
	[C]: in function 'require'
	[string "boot.lua"]:429: in function <[string "boot.lua"]:275>
	[C]: in function 'xpcall'
```

I'm guessing there's a missing default for a preference that you have saved locally. :)

**Comment by fluffy-critter** [comment]
Changing line 73 to:

      if user_settings.grid and user_settings.grid.recent and type(user_settings.grid.recent) == "table" then

fixes it.

**Comment by 25A0** [comment]
Thanks for the quick feedback!

> In the case of grid spilling, that’s actually the behavior I’d expect, since I’m using this for making sprites that are often multiple “cells” large.

That's good to hear; I'll leave it like this then.

> I'm guessing there's a missing default for a preference that you have saved locally. :)

Yeah that's exactly what happened. That should be fixed now, thanks!

**Comment by 25A0** [comment]
Merged into `master` in #32
